java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+ * 's' is our source of randomness. In this case, we can use it directly to get an
+ * integer value.
+ *
+ * Let's assume we have a simple generator for Transactions too. Note that this generator could
+ * build on NonNegativeIntGenerator: both withdrawals and deposits would be expected to be
+ * non-negative. The pattern for this is to pass generators in as constructor parameters
+ * - see ListOf below, or arcs.core.data.PlanParticleGenerator for examples.
+ *
+ * (4) set up a fuzz test using runFuzzTest, your generators, and your invariant. runFuzzTest will
+ * establish a random seed for you, and print the seed value if your tests throws.
+ *
+ * @Test
+ * fuzz_balanceCantGoNegative = runFuzzTest {
+ *   invariant_balanceCantGoNegative(NonNegativeIntGenerator(it), TransactionGenerator(it))
+ * }
+ *
+ * Note that you can also use the invariant for unit tests (the Value class provides an
+ * implementation of 'A' for a single explicit value):
+ *
+ * @Test
+ * aLargeTransactionWontApply() {
+ *   invariant_balanceCantGoNegative(Value(100), Value(listOf(Withdrawal(200))))
+ * }
+ *
+ * It's also easy to use invariant functions to test a range of input conditions:
+ *
+ * @Test
+ * aRangeOfStartingBalances_MaintainInvariant() {
+ *   ...
+ *   listOf(4, 100, 1000, 20, 40, 50).forEach {
+ *     invariant_balanceCantGoNegative(Value(it), transactions)
+ *   }
+ * }
+ *
+ * Finally, you can capture failing fuzz runs explicitly as regression tests:
+ *
+ * @Test
+ * regression_balanceCantGoNegative() {
+ *   var seed = SeededRandom(-35623452)
+ *   invariant_balanceCantGoNegative(NonNegativeIntGenerator(seed), TransactionGenerator(seed))
+ * }
+ *
+ *
+ * A note on types of invariants
+ *
+ * Invariants can take many forms. Here's a very non-exhaustive list:
+ *  - tautologies (e.g. bank account balance is non-negative) - statements that are always true
+ *  - conditional statements (e.g. bank account balance ends up larger if only deposits are
+ *    processed) - statements that are true if a precondition is true
+ *  - differential statements (e.g. St George bank account balances and Commonwealth bank account
+ *    balances end up the same if they start the same and have the same transactions apply)
+ *  - metamorphic statements (if some relation holds between input and output, it still holds after
+ *    certain transformations). Hard to come up with an example using bank accounts, but maybe: if
+ *    the opening balance and magnitude of every transaction is multiplied by a constant factor,
+ *    then so is the closing balance?
+ *
+ * TODO(shanestephens): Add a runRegressionTest function to make reproducing failing fuzz tests
+ * even easier.
+ */
+
+/**
+ * A source of 'T' values. Subclasses may produce the same value every time they're invoked,
+ * or they may provide different values each time.
+ */
+interface A<T> {
+  operator fun invoke(): T
+}
+
+/**
+ * An algorithm that uses an 'I' to generate an 'O' - a Transformer of 'I' into 'O'.
```

Comment:
these letters are really cryptic the first time you read this. I=Input? O=Output?